security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -0,0 +1,241 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sds
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	authapi "github.com/envoyproxy/go-control-plane/envoy/api/v2/auth"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	sds "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/gogo/protobuf/types"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+// SecretType is used for secret discovery service to construct response.
+const SecretType = "type.googleapis.com/envoy.api.v2.Secret"
+
+// TODO(quanlin): secret struct and secretStore interface are placeholders that used for initial check in.
+// will move them to separated file or use existing structure.
+type secret struct {
+	certificateChain []byte
+	privateKey       []byte
+}
+
+type secretStore interface {
+	getSecret() *secret
+}
+
+type discoveryStream interface {
+	Send(*xdsapi.DiscoveryResponse) error
+	Recv() (*xdsapi.DiscoveryRequest, error)
+	grpc.ServerStream
+}
+
+type sdsConnection struct {
+	// PeerAddr is the address of the client envoy, from network layer.
+	PeerAddr string
+
+	// Time of connection, for debugging.
+	Connect time.Time
+
+	// The proxy from which the connection comes from.
+	modelNode *model.Proxy
+
+	// doneChannel will be closed when the client is closed.
+	doneChannel chan int
+
+	// SDS streams implement this interface.
+	stream discoveryStream
+}
+
+type sdsservice struct {
+	st secretStore
+	//TODO(quanlin), add below properties later:
+	//1. workloadRegistry(store proxies information).
+	//2. caClient(interact with CA for CSR).
+}
+
+// newSDSService creates Secret Discovery Service which implements envoy v2 SDS API.
+func newSDSService(st secretStore) *sdsservice {
+	return &sdsservice{
+		st: st,
+	}
+}
+
+// register adds the SDS handle to the grpc server
+func (s *sdsservice) register(rpcs *grpc.Server) {
+	sds.RegisterSecretDiscoveryServiceServer(rpcs, s)
+}
+
+func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecretsServer) error {
+	peerAddr := "Unknown peer address"
+	peerInfo, ok := peer.FromContext(stream.Context())
+	if ok {
+		peerAddr = peerInfo.Addr.String()
+	}
+
+	var discReq *xdsapi.DiscoveryRequest
+	var receiveError error
+	reqChannel := make(chan *xdsapi.DiscoveryRequest, 1)
+
+	con := newSDSConnection(peerAddr, stream)
+	defer close(con.doneChannel)
+
+	go receiveThread(con, reqChannel, &receiveError)
+
+	for {
+		// Block until either a request is received.
+		select {
+		case discReq, ok = <-reqChannel:
```

Comment:
Another case to wait for is:  if secretStore is updated,  push the new secret to the client.   This is the main benefit of gRPC streaming.

Reply:
Yes. But it's a bit subtle in identity provisioning workflow to adopt push model. Because we rely on Envoy SDS to carry the pod credentials in order for node agent update the keycert/secrets.